File: harness/bootstrap_repo_read_execution_contract/manifest.py
"""Caller- and policy-bound expected raw repository file manifest."""

from __future__ import annotations

from typing import Any

from .canonical import ContractError, bounded_canonical_json, self_digest
from .constants import (CANONICALIZATION, MANIFEST_API, MANIFEST_DOMAIN,
                        MAX_MANIFEST_BYTES, MAX_OUTPUT_BYTES, MAX_PATHS)
from .shape import integer, require_keys, sha256, validate_path

FIELDS = {"api_version", "canonicalization", "entries", "kind", "manifest_sha256"}
ENTRY_FIELDS = {"content_bytes", "content_sha256", "kind", "path"}


def manifest_sha256(value: dict[str, Any]) -> str:
    return self_digest(MANIFEST_DOMAIN, value, "manifest_sha256", MAX_MANIFEST_BYTES,
                       "RepoReadExpectedManifest")


def validate_manifest(value: Any) -> dict[str, Any]:
    node = require_keys(value, "RepoReadExpectedManifest", FIELDS)
    bounded_canonical_json(node, MAX_MANIFEST_BYTES, "RepoReadExpectedManifest")
    expected = (MANIFEST_API, CANONICALIZATION, "RepoReadExpectedManifest")
    if (node["api_version"], node["canonicalization"], node["kind"]) != expected:
        raise ContractError("RepoReadExpectedManifest envelope drifted from v1")
    entries = node["entries"]
    if not isinstance(entries, list) or not 1 <= len(entries) <= MAX_PATHS:
        raise ContractError("manifest entries must contain 1..16 exact paths")
    total = _validate_entries(entries)
    if total > MAX_OUTPUT_BYTES:
        raise ContractError("manifest aggregate raw content exceeds 1048576 bytes")
    sha256(node["manifest_sha256"], "manifest.manifest_sha256")
    if node["manifest_sha256"] != manifest_sha256(node):
        raise ContractError("RepoReadExpectedManifest self digest does not match")
    return node


def _validate_entries(entries: list[Any]) -> int:
    paths: list[str] = []
    total = 0
    for index, value in enumerate(entries):
        label = f"manifest.entries[{index}]"
        entry = require_keys(value, label, ENTRY_FIELDS)
        if entry["kind"] != "regular":
            raise ContractError(f"{label}.kind must be 'regular'")
        paths.append(validate_path(entry["path"], f"{label}.path"))
        total += integer(entry["content_bytes"], f"{label}.content_bytes",
                         0, MAX_OUTPUT_BYTES)
        sha256(entry["content_sha256"], f"{label}.content_sha256")
    encoded = [path.encode("utf-8") for path in paths]
    if any(left >= right for left, right in zip(encoded, encoded[1:])):
        raise ContractError("manifest paths must be UTF-8-byte sorted and unique")
    return total


def manifest_paths(value: dict[str, Any]) -> list[str]:
    return [entry["path"] for entry in value["entries"]]


def manifest_content_bytes(value: dict[str, Any]) -> int:
    return sum(entry["content_bytes"] for entry in value["entries"])

File: forge-core/internal/bootstraprepoexecutionauthority/manifest.go
package bootstraprepoexecutionauthority

import "fmt"

var manifestKeys = []string{"api_version", "canonicalization", "entries", "kind", "manifest_sha256"}
var manifestEntryKeys = []string{"content_bytes", "content_sha256", "kind", "path"}

// Manifest is a strict caller-and-policy-bound expected-content precondition.
type Manifest struct{ document map[string]any }

// DecodeManifest validates canonical shape, bounds, ordering, and self digest.
func DecodeManifest(data []byte) (*Manifest, error) {
	document, err := decodeCanonical(data, maxManifestBytes)
	if err != nil {
		return nil, err
	}
	if err = validateManifest(document); err != nil {
		return nil, err
	}
	return &Manifest{document}, nil
}

func (manifest *Manifest) canonicalDocument() map[string]any { return cloneDocument(manifest.document) }

// Entries returns a detached exact read plan in manifest order.
func (manifest *Manifest) Entries() []struct {
	Path          string
	ContentBytes  int64
	ContentSHA256 string
} {
	if manifest == nil {
		return nil
	}
	entries, _ := arrayValue(manifest.document, "entries")
	result := make([]struct {
		Path          string
		ContentBytes  int64
		ContentSHA256 string
	}, 0, len(entries))
	for _, value := range entries {
		entry := value.(map[string]any)
		result = append(result, struct {
			Path          string
			ContentBytes  int64
			ContentSHA256 string
		}{Path: entry["path"].(string), ContentBytes: entry["content_bytes"].(int64),
			ContentSHA256: entry["content_sha256"].(string)})
	}
	return result
}

func validateManifest(document map[string]any) error {
	if err := requireKeys(document, manifestKeys...); err != nil {
		return fmt.Errorf("RepoReadExpectedManifest: %w", err)
	}
	if err := validateEnvelope(document, manifestAPI, "RepoReadExpectedManifest"); err != nil {
		return err
	}
	entries, err := arrayValue(document, "entries")
	if err != nil || len(entries) < 1 || len(entries) > 16 {
		return fmt.Errorf("Manifest entries must contain 1..16 items")
	}
	if err = validateManifestEntries(entries); err != nil {
		return err
	}
	claimed, err := stringValue(document, "manifest_sha256")
	if err != nil || validateHash(claimed, "manifest_sha256") != nil {
		return fmt.Errorf("Manifest self digest is invalid")
	}
	computed, err := selfDigest(manifestDomain, document, "manifest_sha256", maxManifestBytes,
		"RepoReadExpectedManifest", false, "")
	if err != nil || computed != claimed {
		return fmt.Errorf("Manifest self digest does not match")
	}
	return nil
}

func validateManifestEntries(entries []any) error {
	prior, total := "", int64(0)
	for index, value := range entries {
		entry, ok := value.(map[string]any)
		if !ok || requireKeys(entry, manifestEntryKeys...) != nil || entry["kind"] != "regular" {
			return fmt.Errorf("Manifest entry %d shape or kind is invalid", index)
		}
		path, pathErr := stringValue(entry, "path")
		bytes, bytesErr := intValue(entry, "content_bytes")
		if pathErr != nil || validateRepoPath(path) != nil || (index > 0 && prior >= path) ||
			bytesErr != nil || bytes < 0 || bytes > maxContentBytes {
			return fmt.Errorf("Manifest entry %d path, order, or byte count is invalid", index)
		}
		if err := validateHashField(entry, "content_sha256", "Manifest content_sha256"); err != nil {
			return err
		}
		if total > maxContentBytes-bytes {
			return fmt.Errorf("Manifest aggregate content exceeds %d", maxContentBytes)
		}
		prior, total = path, total+bytes
	}
	return nil
}

func validateManifestGrant(manifest *Manifest, grant *issuedGrant) error {
	entries, _ := arrayValue(manifest.document, "entries")
	resources, _ := arrayValue(grant.document, "resources")
	if len(entries) != len(resources) {
		return fmt.Errorf("Manifest path set differs from exact Grant scope")
	}
	for index := range entries {
		entry := entries[index].(map[string]any)
		resource := resources[index].(map[string]any)
		if entry["path"] != resource["path"] {
			return fmt.Errorf("Manifest path set differs from exact Grant scope")
		}
	}
	var total int64
	for _, value := range entries {
		bytes, _ := intValue(value.(map[string]any), "content_bytes")
		total += bytes
	}
	budget := grant.document["budget"].(map[string]any)
	limit, _ := intValue(budget, "max_output_bytes")
	if total > limit {
		return fmt.Errorf("Manifest aggregate bytes exceed Grant output budget")
	}
	return nil
}

func validateManifestAction(manifest *Manifest, actionValue any) error {
	action, ok := actionValue.(map[string]any)
	if !ok {
		return fmt.Errorf("requested_action must be an object")
	}
	resources, _ := arrayValue(action, "resources")
	entries, _ := arrayValue(manifest.document, "entries")
	if len(entries) != len(resources) {
		return fmt.Errorf("Manifest path set differs from requested_action")
	}
	var total int64
	for index, value := range entries {
		entry := value.(map[string]any)
		resource := resources[index].(map[string]any)
		if entry["path"] != resource["path"] {
			return fmt.Errorf("Manifest path set differs from requested_action")
		}
		bytes, _ := intValue(entry, "content_bytes")
		total += bytes
	}
	usage := action["usage"].(map[string]any)
	limit, _ := intValue(usage, "output_bytes")
	if total != limit {
		return fmt.Errorf("Manifest aggregate bytes differ from requested_action output usage")
	}
	return nil
}

func validateEnvelope(document map[string]any, api, kind string) error {
	for key, expected := range map[string]string{"api_version": api,
		"canonicalization": canonicalization, "kind": kind} {
		if err := requireLiteral(document, key, expected); err != nil {
			return err
		}
	}
	return nil
}
